### PR TITLE
AL-112 - Type & Code specific urls for one way sync

### PR DESCRIPTION
**Pull request by jbowdidge-moj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/resource/AlertCodesController.kt [hunk 1]
```diff
@@ -107,6 +109,75 @@ class AlertCodesController(
     httpRequest: HttpServletRequest,
   ) = alertCodeService.deactivateAlertCode(alertCode, httpRequest.alertRequestContext())
 
+  @PreAuthorize("hasAnyRole('$ROLE_ALERTS_ADMIN')")
+  @GetMapping
+  @Operation(
+    summary = "Retrieve an alert code",
+    description = "Retrieve an alert code, typically from the Alerts UI",
```
Comment (bill-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/resource/AlertCodesController.kt:116]: Retrieve _all_ alert codes?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/resource/AlertTypesController.kt [hunk 2]
```diff
@@ -66,6 +66,35 @@ class AlertTypesController(
...
   ): Collection<AlertType> = alertTypeService.getAlertTypes(includeInactive)
 
+  @PreAuthorize("hasAnyRole('$ROLE_ALERTS_READER', '$ROLE_ALERTS_ADMIN', '$PRISON')")
+  @GetMapping("/{alertTypeCode}")
+  @Operation(
+    summary = "Get all alert types",
+    description = "Returns the full list of alert types and the alert codes within those types. " +
+      "By default this endpoint only returns active alert types and codes. " +
+      "The include inactive parameter can be used to return all alert types and codes.",
+  )
```
Comment (bill-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/resource/AlertTypesController.kt:76]: Retrieve an alert type?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/service/AlertCodeService.kt [hunk 3]
```diff
@@ -53,4 +54,13 @@ class AlertCodeService(
       }
     }
   }
+
+  fun retrieveAlertCode(alertCode: String): AlertCode =
+    alertCode.let {
+      it.checkAlertCodeExists()
+      alertCodeRepository.findByCode(it)!!.toAlertCodeModel()
```
Comment (bill-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/service/AlertCodeService.kt:61]: does spring automagically cache value and optimise this, or are we making query to the database 2 times here?

Comment (bill-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/service/AlertCodeService.kt:61]: ok looked it up myself.. restcontroller funcs are transactional by default, hence query results should be automagically cached and reused : )

Comment (jbowdidge-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/service/AlertCodeService.kt:61]: 😄

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/resource/AlertCodesController.kt (hunk 1 above)
Comment (jbowdidge-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/resource/AlertCodesController.kt:116]: Done.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/resource/AlertCodesController.kt [hunk 4]
```diff
@@ -107,6 +109,75 @@ class AlertCodesController(
     httpRequest: HttpServletRequest,
   ) = alertCodeService.deactivateAlertCode(alertCode, httpRequest.alertRequestContext())
 
+  @PreAuthorize("hasAnyRole('$ROLE_ALERTS_ADMIN')")
```
Comment (iterative-it) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/resource/AlertCodesController.kt:112]: I think this is too restrictive? Getting alert types and codes should be available to all so maybe `@PreAuthorize("hasAnyRole('$ROLE_ALERTS_READER', '$ROLE_ALERTS_ADMIN', '$PRISON')")`? The latter is a role all DPS users have

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/resource/AlertCodesController.kt [hunk 5]
```diff
@@ -107,6 +109,75 @@ class AlertCodesController(
...
+        responseCode = "401",
+        description = "Unauthorised, requires a valid Oauth2 token",
+        content = [Content(schema = Schema(implementation = ErrorResponse::class))],
+      ),
+      ApiResponse(
+        responseCode = "403",
+        description = "Forbidden, requires an appropriate role",
+        content = [Content(schema = Schema(implementation = ErrorResponse::class))],
+      ),
+      ApiResponse(
```
Comment (iterative-it) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/resource/AlertCodesController.kt:134]: I don't think this function can return a 404 as it's getting all codes?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/resource/AlertTypesController.kt [hunk 6]
```diff
@@ -66,6 +66,33 @@ class AlertTypesController(
...
+        responseCode = "401",
+        description = "Unauthorised, requires a valid Oauth2 token",
+        content = [Content(schema = Schema(implementation = ErrorResponse::class))],
+      ),
+      ApiResponse(
+        responseCode = "403",
+        description = "Forbidden, requires an appropriate role",
+        content = [Content(schema = Schema(implementation = ErrorResponse::class))],
+      ),
+    ],
```
Comment (iterative-it) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/resource/AlertTypesController.kt:92]: Whereas this one can return a 404 ;)

Comment (jbowdidge-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/resource/AlertTypesController.kt:92]: Done!

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/resource/AlertCodesController.kt (hunk 5 above)
Comment (jbowdidge-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/resource/AlertCodesController.kt:134]: Done.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/resource/AlertCodesController.kt (hunk 4 above)
Comment (jbowdidge-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/resource/AlertCodesController.kt:112]: Done.

**Review by iterative-it — APPROVED** [review]
LGTM

**Pull request merged**